<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

In [1]:
conda install sqlalchemy

Solving environment: done

# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install psycopg2-binary

     |████████████████████████████████| 1.5 MB 750 kB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [63]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [64]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM stores;
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**:  1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT s.name
FROM stores as s
ORDER BY s.name ASC;

"""

df = pd.read_sql_query(sql, engine)
df.head(4)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

**Prompt 2 Answer**:  
3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.category_name, s.vendor, s.btl_price, s.bottle_qty
FROM sales as s
WHERE s.category_name LIKE '%%IMPORTED VODKA%%'
ORDER BY s.vendor ASC
"""

df = pd.read_sql_query(sql, engine)
df.head(3)

category_name                     vendor btl_price  bottle_qty
0  IMPORTED VODKA  Adamba Imports Int'l Inc.     $8.99          12
1  IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           1
2  IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           4

**Prompt 3 Answer**: 


### Prompt 4

What is the total amount (in dollars) of liquor sold in February 2015?

In [60]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)
FROM sales 
WHERE date BETWEEN '2015-02-01' and '2015-02-28';
"""

df = pd.read_sql_query(sql, engine)
df.head()

sum
0  21295350.55

**Prompt 4 Answer**:

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [45]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT s.category_name AS imported_vodka, s.vendor, s.bottle_qty AS amount_sold, s.total
FROM sales as s
WHERE s.category_name LIKE '%%IMPORTED VODKA%%'
ORDER BY s.bottle_qty DESC;

"""

df = pd.read_sql_query(sql, engine)
df

imported_vodka                            vendor  amount_sold  \
0              IMPORTED VODKA   Constellation Wine Company Inc.         1560   
1              IMPORTED VODKA                   Jim Beam Brands         1500   
2              IMPORTED VODKA  Pernod Ricard USA/Austin Nichols         1320   
3              IMPORTED VODKA  Pernod Ricard USA/Austin Nichols         1320   
4              IMPORTED VODKA  Pernod Ricard USA/Austin Nichols         1320   
...                       ...                               ...          ...   
205380         IMPORTED VODKA                   Diageo Americas            1   
205381  IMPORTED VODKA - MISC                           Proximo            1   
205382         IMPORTED VODKA  Pernod Ricard USA/Austin Nichols            1   
205383         IMPORTED VODKA                           Proximo            1   
205384         IMPORTED VODKA   Constellation Wine Company Inc.            1   

           total  
0       34522.80  
1       27180.00  
2       29686.80  
3       29686.80  
4       29686.80  
...          ...  
205380     26.25  
205381     16.76  
205382      9.74  
205383     16.76  
205384      3.81  

[205385 rows x 4 columns]

**Prompt 5 Answer**: Constellation Wine Company Inc sold 1560 bottles totalling $34,522.80 in revenue 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [48]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT s.state_btl_cost - s.btl_price AS Markup
FROM sales AS s

"""

pd.read_sql_query(sql, engine)

markup
0         -$5.75
1         -$5.75
2         -$5.75
3         -$8.58
4         -$2.50
...          ...
3049908   -$5.75
3049909  -$10.00
3049910   -$5.75
3049911   -$5.75
3049912  -$10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup **per store**. Which store has the largest average markup, and what is that average markup?

In [57]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS avg_percentage_markup, s.store
FROM sales s
GROUP BY s.store
ORDER BY avg_percentage_markup DESC
"""

df = pd.read_sql_query(sql, engine)
df.head()

avg_percentage_markup  store
0               0.525471   4013
1               0.521499   4024
2               0.518903   4266
3               0.515094   3833
4               0.514591   3660

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [80]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS avg_percentage_markup, MIN((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS min_percentage_markup, MAX((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS max_percentage_markup, s.store, st.name
FROM sales s
INNER JOIN stores st ON st.store = s.store
GROUP BY s.store, st.name
ORDER BY avg_percentage_markup DESC
"""

df = pd.read_sql_query(sql, engine)
df.head()

avg_percentage_markup  min_percentage_markup  max_percentage_markup  store  \
0               0.525471               0.498571               6.708763   4013   
1               0.521499               0.498462               6.708763   4024   
2               0.518903               0.498462               6.708763   4266   
3               0.515094               0.498462               6.708763   3833   
4               0.514591               0.498462               6.708763   3660   

                         name  
0      Wal-Mart 0841 / Tipton  
1  Wal-Mart 1546 / Iowa Falls  
2  Wal-Mart 1683 / Shenandoah  
3    Wal-Mart 3394 / Atlantic  
4   Wal-Mart 2935 / Knoxville

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold **per county** rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [81]:
# Use a SQL query to find the answer to the above prompt.
# Credit to Haley Taft :)

sql = """
SELECT counties.county, 
       ROUND(AVG(products.bottle_size), 1) AS avg_btl_size,
       ROUND((AVG(sales.bottle_qty)/counties.population), 1) AS avg_per_capita
FROM counties
LEFT JOIN sales
ON counties.county = sales.county
LEFT JOIN products
ON sales.item = products.item_no
GROUP BY counties.county 
"""


df = pd.read_sql_query(sql, engine)
df.head()

county  avg_btl_size  avg_per_capita
0       Wayne        1073.7             0.0
1  Montgomery         997.7             0.0
2   Poweshiek         914.7             0.0
3       Story         921.8             0.0
4     Calhoun        1036.0             0.0

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [ ]:
# -- CAST Syntax:  
# CAST ( expression AS data_type [ ( length ) ] )  
  
# -- CONVERT Syntax:  
# CONVERT ( data_type [ ( length ) ] , expression [ , style ] )  

#    (s.total - (s.btl_price * s.bottle_qty)) as zeroes

In [182]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.btl_price, s.bottle_qty, 
    CAST(s.total AS money) as total, 
    (s.btl_price * s.bottle_qty) as total_1,
    CONVERT(real, (CAST(s.total AS money) - (s.btl_price * s.bottle_qty))) as zeroed
FROM sales as s
"""

# HAVING AVG(zeroed) > 0
df = pd.read_sql_query(sql, engine)
df.head()

ProgrammingError: (psycopg2.errors.UndefinedColumn) column "real" does not exist
LINE 5:     CONVERT(real, (CAST(s.total AS money) - (s.btl_price * s...
                    ^

[SQL: 
SELECT s.btl_price, s.bottle_qty, 
    CAST(s.total AS money) as total, 
    (s.btl_price * s.bottle_qty) as total_1,
    CONVERT(real, (CAST(s.total AS money) - (s.btl_price * s.bottle_qty))) as zeroed
FROM sales as s
]
(Background on this error at: http://sqlalche.me/e/f405)

**Prompt 10 Answer**:

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.btl_price, s.bottle_qty, 
    CAST(s.total AS money) as total, 
    (s.btl_price * s.bottle_qty) as total_1,
    (CAST(s.total AS money) - (s.btl_price * s.bottle_qty)) as zeroed
    
FROM sales as s
"""

# HAVING AVG(zeroed) > 0
df = pd.read_sql_query(sql, engine)
df.head()